com.unity.render-pipelines.universal/ShaderLibrary/Lighting.hlsl

```diff
@@ -567,13 +565,28 @@ half3 VertexLighting(float3 positionWS, half3 normalWS)
 half4 UniversalFragmentPBR(InputData inputData, half3 albedo, half metallic, half3 specular,
     half smoothness, half occlusion, half3 emission, half alpha)
 {
+    // Decreases smoothness as the directionality of the light decreases. This approximates the specular highlight
+    // spreading and scattering, as the light becomes less directional. Without this, even shadowed areas look shiny.
+    // This is recommended by: https://media.contentapi.ea.com/content/dam/eacom/frostbite/files/gdc2018-precomputedgiobalilluminationinfrostbite.pdf
+    // Although, here we do not apply the sqrt to the falloff. Instead we square it, which seems to produce a closer image to the blender groundtruth.
+    // The range remap makes sure that surfaces in perfectly direct light (directionality > .9) remain their true specular (directionality rarely reaches a perfect 1).
+
+    // TODO(john): We could collapse this math down a lot, if we knew the proper transformations from perceptual
+    // smoothness to linear smoothness. Worth optimizing if we become ALU bound. Just make sure to do a before/after.
+    // This smoothness falloff has been carefully tested to look good.
+    half linearSmoothness = 1 - PerceptualSmoothnessToRoughness(smoothness);
```

Comment:
My understanding is that "linear smoothness" is just the term for the "real"/"physical" value. And then 'perceptual smoothness' is just the sqrt(linearSmoothness). This is largely based on similar wordings in other areas: linear color vs perceptual color, where linear again refers to the physical property. 

So in my reading `PerceptualSmoothnessToRoughness` would convert 'perceptual smoothness' to 'linear roughness'. In the end, the only place I actually recall seeing a reference to 'linear smoothness' is the frostbite paper, which terribly does not define it at all. Pixar mentions that squaring the roughness is more 'perceptually linear', but I don't think that's the linear Frostbite refers to when they say 'linear smoothness'. 

That said, you could definitely be right. Implementing the frostbite equation directly didn't result in pleasing values for my testing. So maybe they do intend you to operate on the perceptual smoothness. 

It's all a wash, anyway, because Pixar defines all of their materials with 'perceptual roughness', but Unity uses 'perceptual smoothness'. But `1 - perceptualSmoothness` does not equal perceptual roughness (because it's no longer physical), even though Unity uses it in their conversions. The only accurate conversion is `1-phys_roughness=phys_smoothness` so..
```
1 - roughness_physical = smoothness_physical
sqrt(1 - roughness_physical) = smoothness_perceptual
1 - roughness_physical = smoothness_perceptual^2
roughness_physical = 1 - smoothness_perceptual^2

	so to convert perceptual smoothness to perceptual roughness:
roughness_perceptual = sqrt(1 - smoothness_perceptual^2)
```

Pretty sure that math is correct. Funnily enough, these types of terms totally do affect the specular highlights. Mostly just whether or not the ruler in our scene looks way to shiny. But I don't think it's worth trying to be too close to Pixar, because really what we're trying to match is the PBR renderer of Substance Painter (since lexie is doing all her materials there, haha).

Reply:
I can see how you'd come to that interpretation, but it doesn't correspond to Unity's actual implementation of these terms:
https://github.com/AStrangerGravity/Graphics/blob/asg/master/com.unity.render-pipelines.core/ShaderLibrary/CommonMaterial.hlsl#L42-L45
```
real PerceptualSmoothnessToPerceptualRoughness(real perceptualSmoothness)
{
    return (1.0 - perceptualSmoothness);
}
```

As you say though, it doesn't _really_ matter whether we're using the same "linear smoothness" as the Frostbite paper because it's just a heuristic to make things look subjectively good.

Reply:
Looking over the frostbite paper and considering all of this, it does make me wonder whether they meant 'linearSmoothness' to mean perceptual smoothness. That would explain a lot about why my transformation excludes the sqrt.

Reply:
I've rewritten my code to use physical rather than linear.

Reply:
(Didn't see your comment because Github is archaic)

I think there's two questions here:
  1. What does frostbite mean when they mean linear smoothness? I think I agree they may mean perceptual smoothness.
  2. How do you convert perceptual smoothness to perceptual roughness. I think Unity is actually implementing this incorrectly. It's only valid to subtract from 1, if the values are physical. More precisely, it depends on how Unity and Substance end up calculating perceptual smoothness. It's horrible that there is no standard here.

Reply:
I don't think it matters in the long run because most game engines just don't need to be perfectly physically accurate, and the final look of a game is tuned in situation, anyway.